src/cacheout/cache.py

```diff
@@ -99,6 +104,7 @@ def setup(self) -> None:
         self._cache: OrderedDict = OrderedDict()
         self._expire_times: t.Dict[t.Hashable, T_TTL] = {}
         self._lock = RLock()
+        self._stats_counter = _StatsCounter() if self._stats else None
```

Comment:
I don't want to expose `StatsTracker` for the following reasons:

- If we don't turn on `enable_stats`, `cache.stats.enable()` will cause `'NoneType' object has no attribute 'enable'`. The same goes for other methods.
- It's counterintuitive `Cache` always holds a `StatsTracker` instance to avoid the above problem. If I don't turn on `enable_stats`, why can I get a `StatsTracker` instance?
- We need another lock held by the instance itself if `StatsTracker` is exposed, because the instance may be referenced by other variables.
- Your description seems to combine `StatsTracker` and `Stats` into one class. If I misunderstood, **please ignore this item**. I think `Stats` should be a pure data class, which represents *a snapshot of statistics*. If it has methods such as `enable()` `disable()` and etc, the users may be confused.



So I prefer making `StatsTracker` be an internal instance of `Cache`. The only way to get statistics data is calling a method such as `get_stats()` which return a pure data class `Stats` to get a snapshot of statistics.

If we want to *enable* *disable* *pause* *resume* the `StatsTracker`, We can do it indirectly through `Cache`.

I suggest to change APIs like:

```python
cache.enable_stats() -> None
cache.stats_enabled() -> bool
cache.disable_stats() -> None
cache.reset_stats() -> None
cache.pause_stats() -> None
cache.resume_stats() -> None
cache.get_stats() -> Stats
```

Reply:
1. > If we don't turn on enable_stats, cache.stats.enable() will cause 'NoneType' object has no attribute 'enable'. The same goes for other methods.

`cache.stats` would always be set to an instance of `StatsTracker`.

2. > It's counterintuitive Cache always holds a StatsTracker instance to avoid the above problem. If I don't turn on enable_stats, why can I get a StatsTracker instance?

It would be the main interface for managing/interacting with the stats. If it's disabled, then it just wouldn't be used. Always creating a single instance on the class isn't a concern.

3. > We need another lock held by the instance itself if StatsTracker is exposed, because the instance may be referenced by other variables.

If a lock is needed, then it can be added.

4. > Your description seems to combine StatsTracker and Stats into one class. If I misunderstood, please ignore this item. I think Stats should be a pure data class, which represents a snapshot of statistics. If it has methods such as enable() disable() and etc, the users may be confused.

Sorry if I wasn't clear. Was thinking it would be like you say, two classes, one to manage the stats and one as a pure data class.

5. > So I prefer making StatsTracker be an internal instance of Cache. The only way to get statistics data is calling a method such as get_stats() which return a pure data class Stats to get a snapshot of statistics. If we want to enable disable pause resume the StatsTracker, We can do it indirectly through Cache.

My thinking is to encapsulate the stats interface as a separate thing from `Cache` to have clear separation of concerns and to not have to pollute the interface for `Cache` with whatever stats methods are needed. Another benefit is that it makes the stats tracker easier to override and extend with user defined functionality.